feat: add `string/num-code-points`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add `@stdlib/string/num-code-points` including relevant tests, benchmark and documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   part of  https://github.com/stdlib-js/stdlib/issues/1062 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add @stdlib/string/numCodePoints

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: passed
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented:
I changed 2 or 3 examples in the README.md because they were causing `remark` lint errors something like 'Unknown word' by spell check.

headlessNode commented on lib/node_modules/@stdlib/string/num-code-points/examples/index.js:
```diff
@@ -0,0 +1,33 @@
+
+var numCodePoints = require( './../lib' );
+
+console.log( numCodePoints( 'last man standing' ) );
+// => 17
+
+console.log( numCodePoints( '六书/六書' ) );
+// => 5
+
+console.log( numCodePoints( 'अनुच्छेद' ) );
+// => 8
+
+console.log( numCodePoints( '𐒻𐓟𐒻𐓟' ) );
+// => 4
```
```suggestion
console.log( numCodePoints( 'last man standing' ) );
// => 17

console.log( numCodePoints( '六书/六書' ) );
// => 5

console.log( numCodePoints( '🐶🐮🐷🐰🐸' ) );
// => 5

console.log( numCodePoints( 'Hello 👋 World' ) );
// => 13
```

headlessNode committed:
refactor: match examples

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/string/num-code-points/README.md:
````diff
@@ -0,0 +1,152 @@
+```javascript
+var numCodePoints = require( '@stdlib/string/num-code-points' );
+
+var out = numCodePoints( 'last man standing' );
+// returns 17
+
+out = numCodePoints( '六书/六書' );
+// returns 5
+
+out = numCodePoints( '🐶🐮🐷🐰🐸' );
+// returns 5
+
+out = numCodePoints( 'Hello 👋 World' );
+// returns 13
````
```suggestion
console.log( numCodePoints( 'last man standing' ) );
// => 17

console.log( numCodePoints( '六书/六書' ) );
// => 5

console.log( numCodePoints( '🐶🐮🐷🐰🐸' ) );
// => 5

console.log( numCodePoints( 'Hello 👋 World' ) );
// => 13
```

headlessNode committed:
docs: match examples

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/string/num-code-points/bin/cli:
```diff
@@ -0,0 +1,108 @@
+	args = cli.args();
+
+	// Check if we are receiving data from `stdin`...
+	if ( !stdinStream.isTTY ) {
```
@headlessNode This logic is not what we want. See https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/string/num-grapheme-clusters/bin/cli#L71 and its CLI implementation. That should effectively be replicated here.

kgryte commented on lib/node_modules/@stdlib/string/num-code-points/test/test.cli.js:
```diff
@@ -0,0 +1,261 @@
+		if ( error ) {
+			t.fail( error.message );
+		} else {
+			t.strictEqual( stdout.toString(), '9\n', 'expected value' );
```
This is obviously incorrect. You're supposed to be testing that individual lines are analyzed.

kgryte commented on lib/node_modules/@stdlib/string/num-code-points/test/test.cli.js:
```diff
@@ -0,0 +1,261 @@
+		if ( error ) {
+			t.fail( error.message );
+		} else {
+			t.strictEqual( stdout.toString(), '4\n5\n', 'expected value' );
```
Why are we expecting separate lines here?

kgryte reviewed (changes requested):
Apart from the CLI, this is looking good.

headlessNode committed:
fix: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: passed
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM
